modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/QuantumToolkit.java

```diff
@@ -259,7 +259,25 @@ boolean get() {
          */
         shutdownHook = new Thread("Glass/Prism Shutdown Hook") {
             @Override public void run() {
-                dispose();
+                // Run dispose in a background thread and wait for up to
+                // 5 seconds for it to finish. If it doesn't, then throw an
+                // error, so that if dispose hangs or deadlocks, it won't
+                // prevent the JVM from exiting.
+                var disposeLatch = new CountDownLatch(1);
+                var thr = new Thread(() -> {
+                    dispose();
+                    disposeLatch.countDown();
+                });
+                thr.setDaemon(true);
+                thr.start();
+
+                try {
+                    if (!disposeLatch.await(5, TimeUnit.SECONDS)) {
+                        throw new InternalError("dispose timed out");
```

Comment:
could this happen on a real system?  if so, should the message be more clear and user-friendly?

also, if it can be encountered in the field, it might be better to declare a constant and add a javadoc to it explaining what it is.

what do you think?

Reply:
I think javadoc and a property might be an overkill at this moment. The root problem is the deadlock and is very intermittent, and it hinders with the jenkins jobs. The change seems good means to avoid such situations where process keep running for hours. And if in case the frequency of issue increases or we could find a definite way of reproducing the issue, then we should try to fix the root cause.
Approving, I shall re-approve if updated.

Reply:
I also think that's overkill to document this or make it programmable. This timed wait is to handle the (very rare) case of a hang/deadlock that will otherwise prevent the application from exiting at all. Remember that this code only gets run when the application calls `System.exit(int)`.

I could provide a slightly more informative message. Maybe something like:

```
Timeout shutting down the JavaFX toolkit
```